### ISSUE TITLE
RFC: SemVer for the Docker API

### ISSUE THREAD

**Issue by justincormack** [docker/docker#31842]
When we changed the Docker release versioning in #31075 there were a lot of complaints that we were not using semver http://semver.org/ for the release numbers, or that we had stopped using semver (which was not the case, we have never used it). The part of Docker that is most important for compatibility is the API not the release number, so I want to open a discussion about using semver for that. I have looked at the major points of the semver spec and discussed how they would apply to our release process.

Note this proposal in no way changes the fact that the daemon supports multiple API versions simultaneously. We would continue to support the current 1.x versions as before. We would bump the API to 2.0.0 when Semver is introduced. In addition to all the pre-semver API versions we currently support, each new release will support older minor versions, at some patch version level.

For future backward compatibility, we would need to decide which patch versions we support. For example if the 17.06 release ships with 2.0.0 API, and then we find a bug and we release 17.06.01 with 2.0.1 after 17.07 has shipped with 2.1.0, in most cases we might not forward port the patch bump to the old API support, so 17.07 might just support the original 2.0.0. Alternatively, if the patch was something that also affected 17.07, we could patch bump it to 2.1.1 and patch bump its backport to 2.0.1 at the same time if this was convenient. In general clients SHOULD be written to support any server with the same major version, but if they need particular functionality they MAY require that the server also has a particular minor version. You cannot use the protocol negotiation to request a particular minor version.

This proposal only directly affects the Docker HTTP-JSON API, ie the API from the Docker socket. Other deprecations such as changes in the `Dockerfile` MUST also be included as sending a Dockerfile is part of the supported client communication with the daemon. The Prometheus endpoint is not currently included, but could be added later, or versioned seperately, when it leaves experimental.

Looking at the key points from the semver spec:

_Patch version Z (x.y.Z | x > 0) MUST be incremented if only backwards compatible bug fixes are introduced. A bug fix is defined as an internal change that fixes incorrect behavior._

- These changes would be introduced in major or minor releases as often as needed when bugs are fixed that affect the API.
- Addition of a new field in JSON could be a bug fix if it can be safely ignored, such as the recent case with the stats API, if it is needed to fix a bug. Clients SHOULD ignore unknown fields. If they cannot they will be tied to specific patch releases.
- Changes in http return codes would be a common reason for patch version changes, as now we have a schema for the API rather than just saying it is REST so we should incrememnt if new return codes are introduced.
- If we are currently supporting two releases with the same minor version, we should always increment their patch versions in lockstep by cherry picking the same fixes.
- Once a version of Docker is released we MUST only make patch version changes to its API, not minor or major version changes, as new versions of a release are only bugfixes.

_Minor version Y (x.Y.z | x > 0) MUST be incremented if new, backwards compatible functionality is introduced to the public API. It MUST be incremented if any public API functionality is marked as deprecated. It MAY be incremented if substantial new functionality or improvements are introduced within the private code. It MAY include patch level changes. Patch version MUST be reset to 0 when minor version is incremented._

- We MAY increment this at the three monthly release for the "substantial new functionality" reason. This allows each long term support release to have its own API patch versions, rather than the current interleaving that we have, in case of bug fixes in the API of patch versions. We could increment it more often if needed, if for example a security release causes addition of a new backwards compatible change. We could instead increment the minor version in the monthly after the LTS release to get out of its way, if we have made API changes in the monthly.
- Note the minor version is incremented when a deprecation is announced, so this will be marked in the API docs for that release, even though the functionality cannot be removed until the next major version increment.

_Major version X (X.y.z | X > 0) MUST be incremented if any backwards incompatible changes are introduced to the public API. It MAY include minor and patch level changes. Patch and minor version MUST be reset to 0 when major version is incremented._

- This would happen when we deprecate functionality, or when functionality is changed in a backward incompatible way, such as changing the types of a field.
- Previously we would deprecate functionality on a rolling basis, around nine months after the deprecation announcement. As this now involves a major version change, it would make sense to bunch these, as generally the number of inccompatible changes is small, and deprecation would be the main reason for bumps.
- I suggest bunching deprecations at suitable intervals of no less than one year. When a deprecation is announced, rather than an exact date, we should announce a "no earlier than date" corresponding to the current deprecation period of around nine months, but to decide when to actually deprecate we should look at the urgency, if there are other reasons for a major bump such as incompatible changes, and decide when to increment based on the context.

cc @docker/maintainers @friism @shykes

**Comment by thaJeztah** [comment]
also /cc @ibuildthecloud for input (who did a different proposal in https://github.com/docker/docker/issues/17521)

**Comment by thaJeztah** [comment]
> For future backward compatibility, we would need to decide which patch versions we support. For example if the 17.06 release ships with 2.0.0 API, and then we find a bug and we release 17.06.01 with 2.0.1 after 17.07 has shipped with 2.1.0, in most cases we might not forward port the patch bump to the old API support, so 17.07 might just support the original 2.0.0

The "edge" releases are supported/patched for a month, so chances are "slim" but not _impossible_. If 17.07 is the current "edge" release, it may need to include the patch, depending on the severity of the fix.

**Comment by justincormack** [comment]
@thaJeztah yes, I think in general most API bugs that need a point release would be cherry picked and backported, but if the code has changed significantly due to minor or major bumps it could be ok to not do this under some circumstances.

**Comment by gesellix** [comment]
To put my personal opinion first: I'm neither a friend of SemVer, nor do I trust in any promises SemVer tries to give. This video maybe helps understanding which reasons would oppose the ideas of SemVer: https://www.youtube.com/watch?v=oyLBGkS5ICk (even in case you like SemVer, I suggest watching it, it's quite entertaining ;-) ).

More thoughts about versioning in general: why are we even using version numbers?

You already mentioned one common use case: a client wants to talk to a minimum (or maximum?) api version so that they can expect certain behaviour. SemVer isn't necessary for such a use case, in fact, it makes the code slightly more complex, because you'll need to parse/split/sort dot delimited strings instead of simply using an operator like `<`.

SemVer might allow to "blindly" upgrade your server or client. A nice use case, which is quite popular in the node/npm universe regarding project dependencies. Well... in my experience every new project starts with dynamic version ranges, just to be flexible. But as soon as the project grows and one hits the first breaking change caused by a minor version upgrade you'll start managing your dependencies statically. Just to be safe and not be surprised.

Another use case: I want to perform an api/server upgrade and ensure that it won't break anything. SemVer could communicate such a scenario by increasing the major version. But in fact, this might not have any impact on my individual use case, because my client doesn't even use the broken api endpoint or interface/schema. So, to ensure _my_ client won't be broken, I'll have to read the changelog. I can say that I didn't really care about the Docker version numbers, but I always read the changelog, which helped a lot! You did a good job indicating which endpoints are deprecated, changed or even removed. Maybe there have been some edge cases where a more explicit note was missing, but: if you didn't consider a major change to be noteworthy in a changelog, would you have applied SemVer rules by increasing the major version?

There are other use cases out there, but the last aspect of SemVer versus changelogs is the most important to me. I don't care about the version number. I don't care about the pattern, the promises it tries to make or whatever. I only want to sort version strings or numbers in an easy way so that I can tell if the client or the server is old or new. If I want to know which features are available in which version, I'll read the changelog.

This is why I would accept an api version to be a simple integer. No hidden message, no promises, only the ability to compare between different versions. Other semantics should be put into other places.

Versions are used to communicate that there's a snapshot on a set of features. Just like a git commit hash. Sadly, commit hashes aren't very convenient to work with.
Readmes, changelogs, and log messages are more flexible to express details about changes.
Why mix both aspects?

Sorry for such a long comment, I hope you got the rough idea :-)

**Comment by justincormack** [comment]
@gesellix although I agree linear version numbers are nice and simple, we are maintaining multiple overlapping supported versions, as we have monthly releases, releases supported for four months and releases supported for a year. These need bug fixes, so describing this with a linear number may cause complications. So there is some detail that would need to be thought through with that proposal.

**Comment by dmcgowan** [comment]
I am supportive of this change but think we need to be very explicit about what we consider a "backwards compatible" change. It is fairly obvious when it comes to adding and removing HTTP endpoints or changing JSON structure, but changes to HTTP status codes or even error messages may be incompatible to some clients. Every time an error message is changed there is half a dozen integration tests which needs to be updated accordingly. Clients ideally are not that rigid to the content of the responses, but when they are it can be seen as a backwards incompatible change. In those situations, we are normally take the position that they should not have relied on that string in the first place. Currently our API documentation is not very detailed for responses, does that mean anything not covered in the documentation is subject to change without a corresponding major version change? This is probably most obvious among the image endpoints which do protocol upgrades to TCP then have no documentation for what to expect after the upgrade (https://docs.docker.com/engine/api/v1.26/#operation/ImageCreate). Also, container attach does have some documentation for the format but then it is unclear which endpoints it applies https://docs.docker.com/engine/api/v1.26/#operation/ContainerAttach.

**Comment by justincormack** [comment]
@dmcgowan Yes I think being explicit about the contract is very important. We have a swagger spec now which is encouraging more people to generate clients, which switches a bit more from a REST model to an RPC model. Our return codes are particularly, er, incorrect and attempts to change them have been mixed. I think we should spell out eg that if we change from 4xx to 5xx say then it is a breaking change but 401 to 402 is not (say) if we think that is more useful, but we can say that any return code change is a break if that is preferred.

I don't think having correct documentation is required before something is considered a change, but it obviously makes bugs in eg not bumping major version more likely. Please open issues for underspecified areas.

**Comment by thaJeztah** [comment]
Brought this up in the maintainers meeting, and it looks like everyone was on board, so if you could work on realising this, that'd be great 👍

**Comment by thaJeztah** [comment]
Is this still needed with the plans to move to a gRPC-based API?

**Comment by justincormack** [comment]
Good question... I would say probably yes.
